### PR TITLE
build: Require C++20 compiler

### PR DESCRIPTION
**Pull request by maflcko**
C++20 allows to write safer code, because it allows to enforce more stuff at compile time (`constinit`, `conteval`, `constexpr`, `std::span`, ...).

Also, it allows to write less verbose and easier to understand code (C++ 20 Concepts).

See https://github.com/bitcoin/bitcoin/issues/23363 and https://en.cppreference.com/w/cpp/compiler_support#cpp20

With g++-10 (https://github.com/bitcoin/bitcoin/pull/28348) and clang-13 (https://github.com/bitcoin/bitcoin/pull/28210), there is broad support for almost all features of C++20.

It should be fine to require a C++20 compiler for Bitcoin Core 27.0 in 2024 (next year), not the soon upcoming 26.0 next month.

This pull request includes three small cleanups to make use of C++20 features. If any issues are detected before or after merge, this should be easy to revert. If no issues arise, it should be fine to make use of more involved C++20 features later on.

### CODE REVIEW

File: src/test/fuzz/fuzz.cpp [hunk 1]
```diff
@@ -67,7 +67,7 @@ auto& FuzzTargets()
 
 void FuzzFrameworkRegisterTarget(std::string_view name, TypeTestOneInput target, FuzzTargetOptions opts)
 {
-    const auto it_ins{FuzzTargets().try_emplace(name, FuzzTarget /* temporary can be dropped in C++20 */ {std::move(target), std::move(opts)})};
+    const auto it_ins{FuzzTargets().try_emplace(name, std::move(target), std::move(opts))};
```
[the remark this answers is not in the document]
Comment (fanquake) [reply src/test/fuzz/fuzz.cpp:70]: I assume the problem here is that macOS is going to need a newer version of the SDK for (better) c++20 support

**Comment by hebasto** [comment]
> Not sure why CI fails. Does `./depends/` support C++20?

```diff
--- a/depends/Makefile
+++ b/depends/Makefile
@@ -49,7 +49,7 @@ NO_HARDEN ?=
 FALLBACK_DOWNLOAD_PATH ?= https://bitcoincore.org/depends-sources
 
 C_STANDARD ?= c11
-CXX_STANDARD ?= c++17
+CXX_STANDARD ?= c++20
 
 BUILD = $(shell ./config.guess)
 HOST ?= $(BUILD)
```
?

**Comment by hebasto** [comment]
https://cirrus-ci.com/task/5662608053764096:
> g++: error: unrecognized command line option '-std=c++20'; did you mean '-std=c++2a'?

funny...

**Comment by hebasto** [comment]
> [cirrus-ci.com/task/5662608053764096](https://cirrus-ci.com/task/5662608053764096):
> 
> > g++: error: unrecognized command line option '-std=c++20'; did you mean '-std=c++2a'?
> 
> funny...

It seems we should requires GCC >= 10 first. See: https://gcc.gnu.org/projects/cxx-status.html.

**Comment by ajtowns** [comment]
FWIW, I get a configure.ac error with clang 14 and libstdc++ 13 -- `configure: error: cannot figure out how to use std::atomic`. Seems to be due to clang not correctly implementing `consteval` prior to clang++-15. Using clang's libc++ (`apt-get install libc++abi-dev libc++-dev` and `-stdlib=libc++`) seems like it might fix it.

```
$ head -n31 build-aux/m4/l_atomic.m4  | tail -n19  > testing.cpp
$ clang++ -std=c++17 -stdlib=libc++ -o testing -Wall -W testing.cpp
$ clang++ -std=c++20 -stdlib=libc++ -o testing -Wall -W testing.cpp
$ clang++ -std=c++17 -o testing -Wall -W testing.cpp
$ clang++ -std=c++20 -o testing -Wall -W testing.cpp
In file included from testing.cpp:3:
/usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/chrono:2320:48: error: call to consteval function 'std::chrono::hh_mm_ss::_S_fractional_width' is not a constant expression
        static constexpr unsigned fractional_width = {_S_fractional_width()};
                                                      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/chrono:2320:48: note: undefined function '_S_fractional_width' cannot be used in a constant expression
/usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/chrono:2275:2: note: declared here
        _S_fractional_width()
        ^
1 error generated.
$ clang++-15 -std=c++20 -o testing -Wall -W testing.cpp
$ clang++ --version
Debian clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by ajtowns** [comment]
> If this was the only error, I wonder if it will go away with the cmake transition, or if we can work around it, but I doubt it and expect almost any code to fail to compile.

I think it's specific to `atomic<chrono::duration>` as specified in parts of gcc's libstdc++ headers bracketed by `#if c++20`... Could just decide we're not going to use `atomic<duration>`.

**Comment by ajtowns** [comment]
> > If this was the only error, I wonder if it will go away with the cmake transition, or if we can work around it, but I doubt it and expect almost any code to fail to compile.
> 
> I think it's specific to `atomic<chrono::duration>` as specified in parts of gcc's libstdc++ headers bracketed by `#if c++20`... Could just decide we're not going to use `atomic<duration>`.

Ah, it's worse than that -- clang++-14 can't deal with libstdc++'s chrono header at all in c++20 mode.

```
$ cat testing.cpp
#include <chrono>
int main() { return 0; }
$ clang++-14 -std=c++20 -o testing -Wall -W testing.cpp
In file included from testing.cpp:1:
/usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/chrono:2320:48: error: call to consteval function 'std::chrono::hh_mm_ss::_S_fractional_width' is not a constant expression
        static constexpr unsigned fractional_width = {_S_fractional_width()};
```

Using libc++ with `CC=clang-14 CXX='env clang++-14 -stdlib=libc++' LDFLAGS='-stdlib=libc++'` almost works, but `Db::verify` with clang's ostream isn't available for linking (using system libdb5.3++, not depends build):

```
/usr/bin/ld: libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o): in function `wallet::BerkeleyDatabase::Verify(bilingual_str&)':
./src/./src/wallet/bdb.cpp:327:(.text+0x3367): undefined reference to `Db::verify(char const*, char const*, std::__1::basic_ostream<char, std::__1::char_traits<char> >*, unsigned int)'
```

**Comment by ajtowns** [comment]
> > Using libc++ with `CC=clang-14 CXX='env clang++-14 -stdlib=libc++' LDFLAGS='-stdlib=libc++'` almost works, but `Db::verify` with clang's ostream isn't available for linking (using system libdb5.3++, not depends build):
> 
> Yeah, that should be a known "issue", unrelated to C++20.

Yeah... I guess we could do C++20 with the following options on linux:

 * gcc 10 or above should be fine
 * clang 15 or above should be fine
 * clang 13/14 -- use libc++ and depends build of libdb++

Probably want an autoconf macro to check for the latter case?

**Comment by ajtowns** [comment]
> > Ah, it's worse than that -- clang++-14 can't deal with libstdc++'s chrono header at all in c++20 mode.
> 
> So to clarify, this is only a bug with clang++-14 _and_ libstdc++-13? If so, the only operating system that ships with this config is Debian 13 Trixie, which is your operating system? If so, I'd suggest to just fix it upstream by bumping the default clang version in Debian.

Yes, this seems correct! Seems like a non-problem then.

**Comment by fanquake** [comment]
Our depends version of Qt is also going to be a blocker here. We'll need to move along to a more recent point release for improved C++20 support.

**Comment by hebasto** [comment]
> Our depends version of Qt is also going to be a blocker here. We'll need to move along to a more recent point release for improved C++20 support.

Sorry if I missed something that was mentioned previously, but what is wrong in Qt 5.15.5 with C++20 support?

With this diff:
```diff
diff --git a/depends/Makefile b/depends/Makefile
index 3169117633..319c3498df 100644
--- a/depends/Makefile
+++ b/depends/Makefile
@@ -49,7 +49,7 @@ NO_HARDEN ?=
 FALLBACK_DOWNLOAD_PATH ?= https://bitcoincore.org/depends-sources
 
 C_STANDARD ?= c11
-CXX_STANDARD ?= c++17
+CXX_STANDARD ?= c++20
 
 BUILD = $(shell ./config.guess)
 HOST ?= $(BUILD)
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mk
index 136ce32579..ef30878e65 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -39,7 +39,7 @@ $(package)_config_opts_release += -silent
 $(package)_config_opts_debug = -debug
 $(package)_config_opts_debug += -optimized-tools
 $(package)_config_opts += -bindir $(build_prefix)/bin
-$(package)_config_opts += -c++std c++17
+$(package)_config_opts += -c++std c++2a
 $(package)_config_opts += -confirm-license
 $(package)_config_opts += -hostprefix $(build_prefix)
 $(package)_config_opts += -no-compile-examples
```

everything compiles. Did not test cross-compiling, however.

A quick look at https://github.com/qt/qtbase/compare/v5.15.5-lts-lgpl..v5.15.10-lts-lgpl shows that support for C++23 has been added. But we are talking about C++20, right?

**Comment by fanquake** [comment]
Last I looked there were fixes since 5.15.5 that we needed, and things didn't work. 

> Did not test cross-compiling, however.

What do you mean by "everything"? Probably also worth checking cross-compilation before claiming it works?

**Comment by hebasto** [comment]
> Last I looked there were fixes since 5.15.5 that we needed, and things didn't work.

Why not just sharing known steps to reproduce "things didn't work"?

**Comment by fanquake** [comment]
Because I tested it as part of a different change, months ago, and don't have a trivial reproducer on hand.

**Comment by fanquake** [comment]
> With this diff:

**Comment by fanquake** [comment]
> With this diff:
> everything compiles. 

The first thing I tested after applying your diff failed to compile. Master (fd8ab08558ccfb62f2d0a64d2d0d4dcefba977bc) + #28543 + [the diff above](https://github.com/bitcoin/bitcoin/pull/28349#issuecomment-1743444998):
```bash
Build type: macx-clang (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Compiler: clang (Apple) 15.0.0
Configuration: sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx avx2 avx512f avx512bw avx512cd avx512dq avx512er avx512ifma avx512pf avx512vbmi avx512vl f16c largefile precompile_header rdrnd rdseed shani silent x86SimdAlways release c++11 c++14 c++17 c++1z c++2a reduce_exports static stl
Build options:
  Mode ................................... release
  Optimize release build for size ........ no
  Building shared libraries .............. no
  Using C standard ....................... C11
  Using C++ standard ..................... C++2a
  Using ccache ........................... no
<snip>
compiling ../../corelib/io/qdir.cpp
compiling ../../corelib/io/qdiriterator.cpp
../../corelib/global/qrandom.cpp:386:30: error: no member named 'is_literal_type' in namespace 'std'
        Q_STATIC_ASSERT(std::is_literal_type<SystemAndGlobalGenerators>::value);
                        ~~~~~^
../../../include/QtCore/../../src/corelib/global/qglobal.h:120:57: note: expanded from macro 'Q_STATIC_ASSERT'
#  define Q_STATIC_ASSERT(Condition) static_assert(bool(Condition), #Condition)
                                                        ^~~~~~~~~
../../corelib/global/qrandom.cpp:386:46: error: 'SystemAndGlobalGenerators' does not refer to a value
        Q_STATIC_ASSERT(std::is_literal_type<SystemAndGlobalGenerators>::value);
                                             ^
../../corelib/global/qrandom.cpp:350:26: note: declared here
struct QRandomGenerator::SystemAndGlobalGenerators
                         ^
../../corelib/global/qrandom.cpp:386:74: error: no member named 'value' in the global namespace
        Q_STATIC_ASSERT(std::is_literal_type<SystemAndGlobalGenerators>::value);
                                                                       ~~^
../../../include/QtCore/../../src/corelib/global/qglobal.h:120:57: note: expanded from macro 'Q_STATIC_ASSERT'
#  define Q_STATIC_ASSERT(Condition) static_assert(bool(Condition), #Condition)
                                                        ^~~~~~~~~
compiling ../../corelib/io/qfile.cpp
3 errors generated.
```

Looks like this was fixed upstream in https://github.com/qt/qtbase/commit/df08a21fa4e7139d46ec68bcf264c922789c4f3a, which we don't yet have until updating. Am I missing something?

**Comment by fanquake** [comment]
> Is it required that qt is compiled under C++20 when Bitcoin Core (and other dependencies) are compiled under C++20?

Yes. We've always compiled our dependencies this way, and I can't see a reason to change that now.

File: depends/packages/qt.mk [hunk 2]
```diff
@@ -39,7 +39,7 @@ $(package)_config_opts_release += -silent
 $(package)_config_opts_debug = -debug
 $(package)_config_opts_debug += -optimized-tools
 $(package)_config_opts += -bindir $(build_prefix)/bin
-$(package)_config_opts += -c++std c++17
+$(package)_config_opts += -c++std c++20
```
Comment (hebasto) [inline depends/packages/qt.mk:42]:
```suggestion
$(package)_config_opts += -c++std c++2a
```

**Comment by fanquake** [comment]
If you wanted to see what else will break, could pull in the `std::bit` change: https://github.com/fanquake/bitcoin/commit/67af0da7ee57361860a435615f92c092db060018. Could also bump `assumptions.h` and/or start using `starts_with` in `test/fuzz/rpc.cpp`.

**Comment by fanquake** [comment]
> Not sure what to bump it to. For some reason g++10 has 201709. https://godbolt.org/z/xbY1z4fGM

Hmm. I guess we'll have to just take that value if we want to require c++20, but also support gcc 10. Looking at https://gcc.gnu.org/projects/cxx-status.html, it seems that most language features are implmented by 10.x.

It looks like the value was added to GCC 8, with the addition of c++20 features (https://github.com/gcc-mirror/gcc/commit/026a79f70cf33f836ea5275eda72d4870a3041e5), but then didn't actually get bumped to `202002` until GCC 11 (https://github.com/gcc-mirror/gcc/commit/445430e16bd08ade34637d2346ded40dd49de508), where it was "past time to update this macro".

**Comment by ryanofsky** [comment]
> I wonder if this will be fixed by bumping capnp. Which could make sense independently of the changes here? cc @ryanofsky

I wouldn't say it makes sense independently of the changes here, but bumping would probably be easier than patching. The upstream fix is: https://github.com/capnproto/capnproto/commit/4b35480e18af80b9f0f0c3b9013c2e1ab6d5d7c2 which is pretty recent and only goes back to the 1.0 release, so probably just makes sense to bump to latest 1.01 release

**Comment by fanquake** [comment]
CMake isn't ready yet, and we actually want C++20 first, to save porting over redundant code, just to delete it later.

Happy to take over this pull if you prefer.

**Review by fanquake — APPROVED** [review]
ACK fa6e50d6c79633e22ad4cfc75f56aaa40112ecbb

**Comment by hebasto** [comment]
> Looks like configure doesn't detect g++-10 as a C++20 compiler...

> I presume this will be fixed by cmake...

It [does](https://github.com/hebasto/bitcoin/pull/60) :)

**Comment by fanquake** [comment]
Isn't it just applying the exact same workaround (internally) that we are using? There's no other way to know that GCC supports C++20, because of the bug we discussed above.

**Comment by fanquake** [comment]
Added rel-note todo to the draft wiki.

**Pull request merged**